### PR TITLE
8278050: Armenian text isn't rendered on macOS if text layout is performed

### PR DESCRIPTION
**Pull request by JB-Dmitry**
The problem is related to the implementation of font fallback on macOS, specifically to the path used when text layout
is performed, i.e. the one using `FontSubstitution` and cascade lists. `CTFontCopyDefaultCascadeListForLanguages`'s output contains an entry for the font, able to render Armenian text ('NotoSansArmenian-Regular'), but `CFont.createCompositeFont` cannot find this font amongst registered fonts, and the entry is skipped, resulting in broken rendering.
System font loading (performed by `CFontManager.loadNativeFonts`) relies on `[NSFontManager availableFontFamilies]`, but its output doesn't contain 'Noto Sans Armenian' (which is a family for 'NotoSansArmenian-Regular' font). 'Noto Sans Armenian' is recognized by `[NSFontManager availableMembersOfFontFamily:]` though, 'NotoSansArmenian-Regular' is contained in `[NSFontManager availableFonts]` output, and the font can be created and used normally, just like any other system font. Not sure why its family is 'hidden', maybe macOS developers wanted to limit the number of options
displayed in font selection drop-downs.
For reference, another cascade list item, ignored in similar way is 'NotoSansZawgyi-Regular'.

Proposed fix is to create `CFont` objects for cascade list items explicitly, if corresponding fonts are not registered. These objects are re-used between different fonts as their fallback components (to enable caching of char-to-glyph mapping), but are not made available to public font APIs.
I've considered an alternative solution - using `[NSFontManager availableFonts]` instead of
`[NSFontManager availableFontFamilies]` to load system fonts. Two factors complicate it though - `loadNativeFonts` will
take more time (on my machine about 170 additional fonts will be loaded) and the order of font loading will change (which
impacts the selection of 'preferred' plain/bold/italic/bolditalic components in `FontFamily` class). Both factors can be
dealt with, but they make the solution more complicated and prone to regressions. For now, it seems not worth it.

I've not created a test, as it would rely on macOS providing a fallback font for Armenian script, which is not guaranteed.
Existing tests pass fine with the fix.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278050](https://bugs.openjdk.java.net/browse/JDK-8278050): Armenian text isn't rendered on macOS if text layout is performed

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6633/head:pull/6633` \
`$ git checkout pull/6633`

Update a local copy of the PR: \
`$ git checkout pull/6633` \
`$ git pull https://git.openjdk.java.net/jdk pull/6633/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6633`

View PR using the GUI difftool: \
`$ git pr show -t 6633`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6633.diff">https://git.openjdk.java.net/jdk/pull/6633.diff</a>

</details>

### CODE REVIEW

**Review by prrace — APPROVED** [review]
Looks reasonable. FWIW the rendering being present in the non-layout case is likely due to the magic JRS fall back code .. some day we need to unify this so that at least we are consistent.

**Comment by JB-Dmitry** [comment]
> FWIW the rendering being present in the non-layout case is likely due to the magic JRS fall back code .. some day we need to unify this so that at least we are consistent.

That's what I'm working on now - getting rid of `JRSFontCreateFallbackFontForCharacters` and replace it with cascade list approach. The plan is to have it tested with JetBrains Runtime first (both internally and with external users). If all goes well, I'll submit corresponding PR in several months, after our IDEs' next public release.

**Comment by JB-Dmitry** [comment]
Any other volunteers to review?

**Comment by JB-Dmitry** [comment]
Any chance for a second review?

File: src/java.desktop/macosx/classes/sun/font/CFontManager.java
Comment (mrserb) [inline src/java.desktop/macosx/classes/sun/font/CFontManager.java:337]: I would like to clarify why we cannot merge the results of the availableFontFamilies and availableFonts?

Comment (JB-Dmitry) [reply src/java.desktop/macosx/classes/sun/font/CFontManager.java:337]: > I would like to clarify why we cannot merge the results of the availableFontFamilies and availableFonts?

We can do it, if everyone is OK with increasing the scope of the change, due to the following factors:
* The number of fonts advertised via `GraphicsEnvironment.getAllFonts` and `GraphicsEnvironment.getAvailableFontFamilyNames` will increase, so a potential Java application showing a font chooser combo-box will display more fonts there, than a native macOS application. Not sure whether that's a good or a bad thing, but that's a change in behaviour that will occur.
* As the number of loaded fonts will increase, the execution time of `CFontManager.loadNativeFonts` will also increase, introducing a certain regression in Java UI application startup time. Ideally, this should be fixed simultaneously, e.g. by adjusting `sun.font.FontFamily` logic to perform style resolution lazily (when the particular family is accessed), not at the time `loadNativeFonts` is executed.
* `[NSFontManager availableFonts]` returns the fonts in alphabetical order, while `NSFontManager availableMembersOfFontFamily:]` (used currently) uses font weight to sort the results. The order in which fonts are processed during `loadNativeFonts` impacts the resulting resolution of fonts into styles in `FontFamily`. So, to keep the currently used approach, an additional fonts sorting would need to be introduced - that can be done, e.g. as part of lazy resolution in `FontFamily` mentioned above.

Even though it can make sense to implement all of the above (which JetBrains Runtime already does), I'd rather not do everything in one go. For now, what I'd like to achieve is to switch to `CTFontCopyDefaultCascadeListForLanguages`-based font fallback, as mentioned above. And current change, while having a value of its own, is a necessary step towards that.
If OpenJDK will ever switch to using `[NSFontManager availableFonts]`, as a sole source of system fonts or in addition to `[NSFontManager availableFontFamilies]`, this change can just be reverted.

Comment (mrserb) [reply src/java.desktop/macosx/classes/sun/font/CFontManager.java:337]: Ok, let's investigate that possibility separately.

**Review by mrserb — APPROVED** [review]

File: src/java.desktop/macosx/classes/sun/font/CFontManager.java
Comment (prrace) [reply src/java.desktop/macosx/classes/sun/font/CFontManager.java:337]: I have a plan to re-do a lot of this code. Things like lazily loading the fonts would be good since right now we take a nig it especially right after reboot . But we shouldn't be enumerating (to the API) fonts that are not enumerated in FontBook.
A tricky aspect is that we have had a couple of bugs like this where we are doing fall back essentially using the standard way of doing it that doesn't get as many glyphs as what you get from the JRS code.
Getting away from the JRS code is part of the idea but it needs to be done carefully.

I don't mind you looking into it (obviously not today) but equally I don't promise not to obsolete whatever you do with whatever I do ..

**Pull request closed without merging**
